### PR TITLE
Marimo build, test, deploy workflow

### PR DESCRIPTION
**Pull request by jdangerx**
# Overview

Part of #19 , in conjunction with catalyst-cooperative/eel-hole#116

We need to build & store the webassembly versions of marimo notebooks somewhere, so that we can then grab them in eel-hole (see https://github.com/catalyst-cooperative/eel-hole/pull/116).

What did you change in this PR?
* add stub marimo notebook in marimo/
* update existing workflow to build all `.py` files under `marimo/` to wasm notebooks in `/docs` - and then deploy to GH pages
  * write a lil python script to do the grunt work here since this got gnarly in bash
  * if we want to make [islands](https://docs.marimo.io/guides/island_example/) later we can do that
  * if we want to deploy the ipynbs also via nbconvert etc, we can do that in a different workflow, or add a different script to this workflow and rename it from marimo-only. All out of scope for now though.

Necessary chores to make this work:
* add `uv` (we *could* also use `pixi` but so far seems unnecessary? easy enough to swap though)
* set up pre-commit so we stop getting that annoying ci error.

# Testing

* ran action manually, downloaded [build artifact](https://github.com/catalyst-cooperative/pudl-examples/actions/runs/22321377400/artifacts/5623059193), served `python3 -m http.server` from artifact.
* found that you have to merge to `main` before we can deploy to gh-pages
# To-do list

- [ ] Once approved, commit results of `uv run pre-commit run -a` to this branch so we get a nice green check mark, but don't clutter up these changes with those changes.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+requires-python = ">=3.13"
+dependencies = [
+    "fastparquet>=2024.11.0",
+    "geopandas>=1.1.2",
+    "jupyter>=1.1.1",
+    "marimo>=0.19.11",
+    "matplotx>=0.3.10",
+    "numpy>=2.4.2",
+    "pandas>=2.3.2",
+    "pyarrow>=23.0.1",
```
Comment (zaneselvans) [inline pyproject.toml]: Kaggle is trapped on goepandas 0.14, which also limits the pyarrow version that can be installed, so this environment will be incompatible with the environment that's currently used to run the Jupyter notebooks on Kaggle. Not sure what to do about that, but it might result in different behavior.

File: README.md [hunk 2]
```diff
@@ -24,36 +31,42 @@ any data.
...
+```
+$ uv run jupyter lab
+```
+
+To run the marimo notebooks, navigate to the `marimo` subdirectory, then run:
 
-If you're already familiar with git, Python environments, filesystem paths, and running
-Jupyter notebooks locally, you can also work with these notebooks and the PUDL data locally:
+```
+$ uv run marimo edit
```
[the remark this answers is not in the document]
Comment (zaneselvans) [reply README.md]: To avoid this kind of oops, did you find an easy way to try and run all the notebooks in CI? I guess the export to WASM just translates them and doesn't execute them, right?

**Review by zaneselvans — CHANGES_REQUESTED** [review]
I ran into errors when attempting to execute both the Jupyter and Marimo notebooks locally because of dependency stuff. See below.

File: scripts/export_marimo_notebooks.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    if not notebooks:
+        raise RuntimeError(f"No marimo notebooks found under {source_dir}/.")
+
+    for notebook in notebooks:
+        relative_path = notebook.relative_to(source_dir)
+        output_path = output_dir / relative_path.with_suffix(".html")
+        output_path.parent.mkdir(parents=True, exist_ok=True)
+
+        print(f"Exporting {notebook} to {output_path}")
+        subprocess.run(
```
[the remark this answers is not in the document]
Comment (jdangerx) [reply scripts/export_marimo_notebooks.py]: Don't think so. There's one for [islands](https://docs.marimo.io/guides/exporting/#generating-islands) which might be interesting down the line.

File: tests/playwright/conftest.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            with urlopen(url):
+                return
+        except URLError:
+            time.sleep(sleep_seconds)
+    raise RuntimeError(f"Timed out waiting for test server at {url}.")
+
+
+@pytest.fixture(scope="session")
+def server_bind() -> str:
+    return os.environ.get("PUDL_TEST_SERVER_BIND", "localhost:8001")
```
Comment (jdangerx) [inline tests/playwright/conftest.py]: If you have multiple test processes running at once this will get mad at you... but I didn't really want to gunk this up with "find a random unused port" logic for that edge case.
